BibtexQuery/Format.lean

```diff
@@ -170,133 +170,154 @@ partial def getLastCharOfContent (x : Content) : Option Char :=
-  if arr.isEmpty then .none else .some arr
+def words (arr : Array (Option (Array Content))) : Array Content :=
+  arrayConcat arr #[.Character " "]
 
-def sentence1 (arr : Array Content) : Option (Array Content) :=
+def canEndSentence (c : Char) : Bool :=
+  match c with
+  | '.' | '?' | '!' | '…' | '。' | '？' | '！' | '\uFF0E' | '\uFF61'
+  | '\u203C' | '\u203D' | '\u2047' | '\u2048' | '\u2049' => true
+  | _ => false
+
+def sentence1 (arr : Array Content) : Array Content :=
   if arr.isEmpty then
-    .none
-  else if GeneralCategory.isPunctuation (getLastCharOfArrayContent arr |>.getD default) then
-    .some arr
+    #[]
+  else if canEndSentence (getLastCharOfArrayContent arr |>.getD ' ') then
+    arr
   else
-    .some (arr ++ #[.Character "."])
+    arr ++ #[.Character "."]
 
-def sentence (arr : Array (Option (Array Content))) : Option (Array Content) :=
+def sentence (arr : Array (Option (Array Content))) : Array Content :=
   sentence1 <| arrayConcat arr #[.Character ", "]
 
-def sentence' (s : String) : Array Content :=
-  if s.isEmpty then
-    #[]
-  else if GeneralCategory.isPunctuation s.back then
-    #[.Character s]
-  else
-    #[.Character (s ++ ".")]
+def sentence' (content : Array TexContent) : Array Content :=
+  sentence1 <| TexContent.toHtmlArray content
 
 def toplevel (arr : Array (Option (Array Content))) : Array Content :=
   arrayConcat arr #[.Character "\n"]
 
+/-!
+
+### Partial templates
+
+-/
+
 def formatVolumeAndPages : Option (Array Content) := do
   let pages ← e.tags.find? "pages"
   match e.tags.find? "volume" with
   | .some volume =>
     arrayConcat #[
-      mkStr' volume,
+      mkStr volume,
       (mkStr · "(" ")") <$> e.tags.find? "number",
-      mkStr' pages ":"
+      mkStr pages ":"
     ]
-  | .none => mkStr' pages "pages "
+  | .none => mkStr pages "pages "
 
-def formatWebRefs : Option (Array Content) :=
+def formatWebRefs : Array Content :=
   let formatUrl : Option (Array Content) := do
-    let s ← e.tags.find? "url"
-    pure <| if s.isEmpty then #[] else #[.Character "URL: ", mkHref s s]
+    let s ← TexContent.toPlaintextArray <$> e.tags.find? "url"
+    if s.isEmpty then .none else .some <| #[.Character "URL: "] ++ mkHref s #[.normal s]
   let formatUrl' (tagName urlPrefix namePrefix : String) : Option (Array Content) := do
-    let s ← e.tags.find? tagName
-    pure <| mkHref' (urlPrefix ++ s) (namePrefix ++ s)
+    let s ← TexContent.toPlaintextArray <$> e.tags.find? tagName
+    let s :=
+      if s.startsWith urlPrefix then
+        s.drop urlPrefix.length
+      else if s.startsWith namePrefix then
+        s.drop namePrefix.length
+      else
+        s
+    if s.isEmpty then .none else mkHref (urlPrefix ++ s) #[.normal (namePrefix ++ s)]
   sentence #[
     formatUrl' "pubmed" "https://www.ncbi.nlm.nih.gov/pubmed/" "PMID:",
     formatUrl' "doi" "https://doi.org/" "doi:",
     formatUrl' "eprint" "https://arxiv.org/abs/" "arXiv:",
     formatUrl
```

Comment:
Do I understand correctly that this code assumes that all `eprint` fields refer to the arXiv, regardless of what is in the `eprinttype` field?

Reply:
> that this code assumes that all `eprint` fields refer to the arXiv

Unfortunately, yes. That's the code I found in `pybtex`, it processes arxiv, alongside with pubmed and doi.

Do you have documents on `eprinttype`?

PS: I'm on vacation on August, so it's possible that I can't reply in time.

Reply:
I can't find any official specs about it, but lots of people are using it for multiple online archiving services using custom code. (In fact I do this with the IACR eprint archive, for instance.)

Reply:
The only three pages I found on google mentioning `eprinttype` are:
- https://tex.stackexchange.com/questions/542600/how-to-use-eprinttype-with-medrxiv
- https://www.overleaf.com/learn/latex/Bibliography_management_with_biblatex
- https://en.wikibooks.org/wiki/LaTeX/Bibliography_Management

So I'd like to ask for some examples, e.g. a list of `eprinttype = ??? --> URL: https://??? and title: ???`

Reply:
> My understanding is that it's meant to be user-extensible. For us right now, I think it would be fine to only support the arXiv, so if `eprinttype` is `arXiv`, then add the arxiv link, otherwise do nothing.

I think it's better defaults to arXiv if `eprinttype` is not set. By the way, in https://github.com/leanprover-community/mathlib4/blob/master/docs/references.bib not all of entries have `eprinttype`. Some of them uses `archiveprefix` instead. Some of them only have `eprint` and have no other fields. Some of them misuse `eprint` as `url`.

Reply:
Alright, fair enough.